File: engine/src/schemas/strategy.py
"""Pydantic models for Strategy entity."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class StrategyCreate(BaseModel):
    """Request body for creating a new strategy."""
    name: str = Field(min_length=1, max_length=200)
    description: str | None = None
    source_code: str = Field(min_length=1)
    parameters: dict[str, Any] = Field(default_factory=dict)


class StrategyResponse(BaseModel):
    """Full strategy entity response."""
    id: UUID
    version_hash: str = Field(pattern=r"^sha256:[a-f0-9]{64}$")
    name: str
    description: str | None = None
    source_code: str
    parameters: dict[str, Any]
    created_at: datetime
    created_by_session: str | None = None
    parent_version_hash: str | None = None


class StrategyListItem(BaseModel):
    """Strategy summary for list endpoints."""
    id: UUID
    version_hash: str
    name: str
    created_at: datetime
    backtest_count: int = 0
    live_session_count: int = 0


class StrategyListResponse(BaseModel):
    """Paginated strategy list response."""
    items: list[StrategyListItem]
    total: int


class StrategyCreateResponse(BaseModel):
    """Response after creating a strategy."""
    id: UUID
    version_hash: str
    name: str
    created_at: datetime

File: packages/shared/src/schemas/strategy.ts
import { z } from 'zod';

export const StrategyParametersSchema = z.record(z.string(), z.unknown());

export const StrategyCreateSchema = z.object({
  name: z.string().min(1).max(200),
  description: z.string().optional(),
  source_code: z.string().min(1),
  parameters: StrategyParametersSchema.optional().default({}),
});

export const StrategySchema = z.object({
  id: z.string().uuid(),
  version_hash: z.string().regex(/^sha256:[a-f0-9]{64}$/),
  name: z.string(),
  description: z.string().optional(),
  source_code: z.string(),
  parameters: StrategyParametersSchema,
  created_at: z.string().datetime(),
  created_by_session: z.string().optional(),
  parent_version_hash: z.string().regex(/^sha256:[a-f0-9]{64}$/).nullable().optional(),
});

export const StrategyListItemSchema = z.object({
  id: z.string().uuid(),
  version_hash: z.string(),
  name: z.string(),
  created_at: z.string().datetime(),
  backtest_count: z.number().int().nonnegative(),
  live_session_count: z.number().int().nonnegative(),
});

export const StrategyListResponseSchema = z.object({
  items: z.array(StrategyListItemSchema),
  total: z.number().int().nonnegative(),
});

export const StrategyCreateResponseSchema = z.object({
  id: z.string().uuid(),
  version_hash: z.string(),
  name: z.string(),
  created_at: z.string().datetime(),
});

export type Strategy = z.infer<typeof StrategySchema>;
export type StrategyCreate = z.infer<typeof StrategyCreateSchema>;
export type StrategyListItem = z.infer<typeof StrategyListItemSchema>;
export type StrategyListResponse = z.infer<typeof StrategyListResponseSchema>;
export type StrategyCreateResponse = z.infer<typeof StrategyCreateResponseSchema>;
